samcli/lib/utils/arn_utils.py

```diff
@@ -31,11 +32,47 @@ class ARNParts:
     service: str
     region: str
     account_id: str
+    resource_type: str
     resource_id: str
 
     def __init__(self, arn: str) -> None:
-        parts = arn.split(":")
         try:
-            [_, self.partition, self.service, self.region, self.account_id, self.resource_id] = parts
+            # Regex pattern formed based on the 3 ARN general formats found here:
+            # https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html
+            arn_pattern = (
+                r"arn:([a-zA-Z0-9_-]+):"  # Pattern for partition
+                r"([a-zA-Z0-9_-]+):"  # Pattern for service
+                r"([a-zA-Z0-9_-]*):"  # Pattern for region
+                r"([a-zA-Z0-9_-]*)"  # Pattern for account_id
+                r"(?::([a-zA-Z0-9_-]+))?"  # Pattern for resource_type if it exists
+                r"(?::(.+))"  # Pattern for resource_id if it exists
+            )
+            print(arn_pattern)
+            matched_arn = re.match(arn_pattern, arn)
+            if not matched_arn:
+                raise ValueError()
+
+            self.partition = matched_arn.group(1)
+            self.service = matched_arn.group(2)
+            self.region = matched_arn.group(3) if matched_arn.group(3) else ""
+            self.account_id = matched_arn.group(4) if matched_arn.group(4) else ""
+            self.resource_type = matched_arn.group(5) if matched_arn.group(5) else ""
+            if matched_arn.group(5):
+                # This handles the Arns of services with the format:
+                # arn:partition:service:region:account-id:resource-type:resource-id
+                self.resource_type = matched_arn.group(5)
+                self.resource_id = matched_arn.group(6) if matched_arn.group(6) else ""
+            elif "/" in matched_arn.group(6):
+                # This handles the Arns of services with the format:
+                # arn:partition:service:region:account-id:resource-type/resource-id
+                split_resource_type_and_id = matched_arn.group(6).split("/", 1)
+                self.resource_type = split_resource_type_and_id[0]
+                self.resource_id = split_resource_type_and_id[1]
+            else:
+                # This handles the Arns of services with the format:
+                # arn:partition:service:region:account-id:resource-id
+                self.resource_type = ""
+                self.resource_id = matched_arn.group(6) if matched_arn.group(6) else ""
+
         except ValueError as ex:
```

Comment:
Is this exception valid anymore, or should this be replaced something else? I am not sure what will happen if a random string is passed here?

Reply:
That's True, we can raise the `InvalidArnValue` on line 53 instead and remove line the try/except here. This will be caught in remote_invoke_context.py and a `UserException` will be thrown with the message `Can't find exact resource information with given {self._resource_id}. "
f"Please provide full resource ARN or --stack-name to resolve the ambiguity.`